WIP: Upgrade Clue to PHP 7.3 and Laravel 6.5

chrisrlc opened the pull request:
**Marking as Draft/Not Ready to Merge because once this PR looks good to go, we still want to wait for a few more things to finish up on our end before merging to master and deploying to production.**

Changelog (biggest changes marked in **bold**):

*  Upgraded laravel/framework from 5.4 to 6.5
    - [5.8 to 6.x](https://laravel.com/docs/6.x/upgrade):
         - Added laravel/helpers package to composer.json to continue using deprecated array_ and str_ helpers
         - Use correct param keys when generating urls in views (i.e. admin/team/645/edit, not admin/team//edit?id=645)
    - [5.6 to 5.7](https://laravel.com/docs/5.7/upgrade):
         - Switched to php "??" instead of blade "or" in views
         - Added a new storage/framework/cache/data directory
         - Added MAIL_LOG_CHANNEL option to config/mail.php
    - [5.5 to 5.6](https://laravel.com/docs/5.6/upgrade):
         - Added config/logging.php and config/hashing.php
    - [5.4 to 5.5](https://laravel.com/docs/5.5/upgrade):
        - **Switched to using Swift Mailer included in Laravel 5.5 instead of phpmailer**
             - removes default Swift Mailer email styling
         - **Uses Laravel's Parsedown dependency for creating markdown email and parsing markdown in views** (instead of andreasindal/laravel-markdown)
         - Added package:discover to post-autoload-dump
         - php artisan optimize no longer needed (and deprecated in laravel 5.6)
         - Auto-load commands in App\Console\Kernel
* [Upgraded adldap2/adldap2-laravel from 3.0 to 6.0](https://adldap2.github.io/Adldap2-Laravel/#/upgrading)
    - Renamed everything from adldap to ldap
    - Adds auth event logging (turned off by default)
    - Uses updated ldap and ldap_auth configs (mostly just renamed vars)
    - Added objectguid db column to players and agents tables for proper data sync
* Upgraded laravel-breadcrumbs from 3.0 to 5.3
    - Remove from list of app providers and facades
    - Replaced Breadcrumbs::for with Breadcrumbs::register
    - Replaced $breadcrumbs with $trail
* Upgraded npm devDependencies
    - **Switched to using Laravel's asset compiler laravel-mix w/ webpack, instead of laravel-elixir w/ gulp** (laravel-elixir is no longer supported and relies on an older version of gulp)
         - Relies on new webpack.mix.js instead of gulpfile.js
         - Loads vendor fonts in public/fonts/vendor
         - Loads custom fonts in public/fonts instead of public/fonts/clueicons or public/fonts/socialicons
               - Note: I didn't actually want to change the custom fonts directory structure, but couldn't figure out a way to keep this structure without completely overriding Laravel Mix/webpack's [URL processing](https://laravel.com/docs/6.x/mix#url-processing). Very open to suggestions here!
         - Compiles resources/assets/admin.scss to public/ (can only compile to public/) and concatenates this w/ datepicker css to all-admin.css
    - Upgraded jquery, clipboard
    - Added [Laravel's node package dependencies and scripts](https://github.com/laravel/laravel/blob/master/package.json): axios, cross-env, lodash, resolve-url-loader, sass, sass-loader
* Upgraded PHP from 5.6 to 7.3
    - Set previously undefined variables in compact(): [Before PHP 7.3, any strings that are not set will silently be skipped.](https://www.php.net/manual/en/function.compact.php)
* Configured Laravel log to be able to write to /logs/laravel directory

queryluke commented on app/Mail/Registered.php:
```diff
@@ -0,0 +1,46 @@
+        unset($pieces[0]);
+        $body = implode('', $pieces);
+
+        return $this->markdown('emails.registered', ['body' => $body])->subject($subject);
```
Is `$this->markdown()` using Parsedown?

chrisrlc replied:
It is! Laravel 5.5+ comes with Parsedown for writing markdown messages. I followed the [Laravel guide for generating markdown mail](https://laravel.com/docs/6.x/mail#markdown-mailables) and [this Medium article](https://medium.com/@DarkGhostHunter/laravel-there-is-a-markdown-parser-and-you-dont-know-it-5f523e22121e) for using Laravel's Parsedown parsing markdown elsewhere.

queryluke replied:
great, just checking

queryluke commented on database/migrations/2019_11_20_155351_add_objectguid_column.php:
```diff
@@ -0,0 +1,40 @@
+    public function up()
+    {
+        Schema::table('agents', function (Blueprint $table) {
+            $table->string('objectguid')->unique()->nullable()->after('id');
```
What's the purpose of `objectguid`? Just so I know for myself.

chrisrlc replied:
The objectguid is a new column required in the adldap2-laravel package upgrade from 5.* to 6.* when using the DatabaseUserProvider: https://adldap2.github.io/Adldap2-Laravel/#/upgrading?id=authentication. It's supposed to "allow usernames to change in your directory and synchronize properly in your database".

queryluke replied:
Cool, I did experience an issue when trying to run the migration locally:

```
In StreamHandler.php line 98:
                                                                               
  Missing stream url, the stream can not be opened. This may be caused by a premature call to close().
```

It did run other migrations though, so I don't think it's my mysql connection. Have you have any issues?

queryluke commented on readme.md:
````diff
@@ -50,27 +50,34 @@ $ ln -nfs $CLUE_HOME/storage/app/public $CLUE_HOME/public/storage
-$ gulp watch
-```
+   ```bash
+   $ cd $MRBS_HOME
````
Check `$MRBS_HOME`, should be `$CLUE_HOME`, correct? Probably copy-pasta from upgrading MRBS

chrisrlc replied:
Yikes, fixed. Thanks for catching!

queryluke commented:
Looking good! Thanks for fixing all my stupid typos 🤦‍♂ 

I've reviewed the code changes and left a few comments. I need to reinstall Clue on my home computer and test the changes. Will comment when that's complete.

chrisrlc commented on resources/assets/sass/ui.scss:
```diff
@@ -154,10 +154,11 @@ legend {
-  height: 175px;
-  width: 175px;
-}
+// TODO: What element should this be for?
```
@queryluke This media query broke the compiler because it wasn't referring to a css element - do you know what it should be referring to, or should I just remove this?

queryluke replied:
oh, that's odd. well if it's not referencing a css element, then it's not doing anything. so yeah I'd call it safe to delete.

chrisrlc replied:
Deleted.

chrisrlc commented:
> Saw that admin.css, ui.css, and web.css were removed. Are these compiled elsewhere, or did you compile everything into all-admin.css?

admin.css was previously compiled by gulp from resources/assets/sass into resources/assets/css, then concatenated with eonasdan-bootstrap-datetimepicker css into admin.css under public/. Webpack seemed to only want to compile from resources into public only, so I had to move that first compiled admin.css to public/css (and hence re-titling old admin.css to all-admin.css).

ui.css and web.css should have been left where they were in public/ ? The contents changed a bit because I had to update font paths, but otherwise, I believe they were mostly left alone.

chrisrlc commented:
Thanks so much for reviewing thus far! I made 3 additional changes based on your comments:

1. fixed a typo in the readme
2. removed old adldap configs that I accidentally left in (they're unused and were replaced by config/ldap.php and config/ldap_auth.php)
3. removed unused public/resources/assets/css/admin.css that I must have accidentally generated with webpack when I was first trying to get it set up

queryluke commented:
Sorry for the delay. I'm finally set up locally. Can't get the fake player seeder to run, but I can figure that out later. It was only for my own personal development.

I'll do some actual interface testing this week.

queryluke commented:
Ok! this looks ready to go! I checked all the routes and controller functions on my local machine and everything is working well (just one minor issue with a redirect which I pushed to this branch).

Really well done with the upgrade! 👍

chrisrlc commented:
Thanks so much for reviewing @queryluke! We'll merge to master on our end when we're ready to get this on production.

chrisrlc commented:
@queryluke - Just FYI that I made a small commit to fix a couple typos that were discovered during QA testing.

queryluke commented:
Of course, I'm sure there are more typos around. Thanks for fixing them.

queryluke commented:
@chrisrlc 
Christina, I caught this while helping McGill set up their clue instance. I've got an errand die-dump `dd()` here: https://github.com/UNC-Libraries/wilson-clue-app/blob/master/app/Http/Middleware/Admin.php#L15

It's not a huge deal, but I figured you could included it in this big PR.
